runtime/src/identity.rs

```diff
@@ -380,6 +396,136 @@ decl_module! {
             Ok(())
         }
 
+        /// Initiates master key rotation for a DID.
+        ///
+        /// # Failure
+        /// Only called by master key owner.
+        ///
+        /// # Arguments
+        /// * `did` caller's DID
+        /// * `new_key` master key replacing current key
+        pub fn change_master_key(origin, did: IdentityId, new_key: Key) -> Result {
+            let sender = ensure_signed(origin)?;
+            let master_key = Key::try_from(sender.encode())?;
+            let _grants_checked = Self::grant_check_only_master_key(&master_key, did)?;
+
+            ensure!(!<MasterKeyRotation>::exists(did), "a key rotation process already in progress");
+
+            let rotation = MasterKeyRotationPrefs {
+                new_key,
+                target_accepted: false,
+                kyc_verified: false
+            };
+
+            <MasterKeyRotation>::insert(did, rotation);
+
+            Self::add_auth(Signer::Key(master_key), Signer::Key(new_key), AuthorizationData::RotateMasterKey(master_key, did), None);
+
+            Self::deposit_event(RawEvent::MasterKeyChangeInitiated(did, new_key));
+
+            Ok(())
+        }
+
+        /// Call this with the new master key. By invoking this method, caller accepts authorization
+        /// `auth_id` with the new master key. If a KYC service provider approved this change, master
+        /// key of the DidRecord is updated
+        ///
+        /// # Arguments
+        /// * `did` caller's DID
+        /// * `auth_id` AuthorizationID that corresponds to changing master key
+        pub fn accept_master_key(
+            origin,
+            did: IdentityId,
+            auth_id: u64
+        ) -> Result {
+            let new_key = Key::try_from(ensure_signed(origin)?.encode())?;
+            let new_signer = Signer::from(new_key);
+
+            ensure!(<Authorizations<T>>::exists((new_signer, auth_id)), "Invalid auth");
+            let auth = Self::authorizations((new_signer, auth_id));
+
+            let (current_master_key, did) = match auth.authorization_data {
```

Comment:
`current_master_key` must be fetched directly from the Identity since anyone can set anything in the authorizations.